### PR TITLE
libexpr: deprecate using "or" as a name

### PR DESCRIPTION
**Pull request by regnat**
For historical reasons, "or" (which is a keyword) can also be used as a
variable or an attribute name, which requires some weird special cases
in the parser.

This pr adds a warning whenever one of those are used, with the hope
that this special case will be removed.

Related to NixOS/nixpkgs#25475

### CODE REVIEW

**Comment by edolstra** [comment]
I think it should be "deprecate" rather than "depreciate".

The warning should use `warnOnce` to prevent spamming the user over and over.

**Comment by regnat** [comment]
> I think it should be "deprecate" rather than "depreciate".

Oops, my bad

> The warning should use |warnOnce| to prevent spamming the user over and
> over.

Thanks, fixed

File: src/libexpr/parser.y [hunk 1]
```diff
@@ -483,7 +490,13 @@ attrpath
 
 attr
   : ID { $$ = $1; }
-  | OR_KW { $$ = "or"; }
+  | OR_KW {
+    warnOnce(
+      haveWarnedAboutOrVar,
+      format("using \"or\" as an attribute name is depreciated at %1%")
```
Comment (veprbl) [inline src/libexpr/parser.y]: s/depreciated/deprecated/

File: src/libexpr/parser.y [hunk 2]
```diff
@@ -358,7 +359,13 @@ expr_select
   | /* Backwards compatibility: because Nixpkgs has a rarely used
        function named ‘or’, allow stuff like ‘map or [...]’. */
     expr_simple OR_KW
-    { $$ = new ExprApp(CUR_POS, $1, new ExprVar(CUR_POS, data->symbols.create("or"))); }
+    {
+      warnOnce(
+        haveWarnedAboutOrVar,
+        format("using \"or\" as a variable is depreciated at %1%")
```
Comment (veprbl) [inline src/libexpr/parser.y]: s/depreciated/deprecated/

**Comment by copumpkin** [comment]
I'm on the fence about that `warnOnce` stuff. On one hand, I don't want to get spammed, but on the other, I don't want warnings to trickle out one at a time when my goal is to fix them. Take a project with a ton of `or` identifiers and `or` keywords. I want to convert all the identifiers, so I'd like to get a big list o' failures, but `warnOnce` hides all but the first each time, so I need to re-evaluate after each build.

This is a broader concern, but we might as well stay consistent here.

**Comment by regnat** [comment]
@copumpkin I agree with this, probably adding a cli flag to make `warnOnce` display all warnings instead of just one could bring the best of both approaches.

**Pull request closed without merging**